### PR TITLE
std::optional

### PR DESCRIPTION
**Pull request by TurkeyMan**
I think this is done; just gotta get the tests in order...

### CODE REVIEW

**Comment by thewilsonator** [comment]
LMK when its good.

**Comment by TurkeyMan** [comment]
It's not good. Unit tests never work, and there's no way to build and run tests locally, and I just CBF.

**Comment by 12345swordy** [comment]
You going to finished up std::vector hopefully?

**Comment by TurkeyMan** [comment]
`std::vector` works on windows.

**Comment by TurkeyMan** [comment]
Help make the unit tests work and I'll do the work to support the other platforms. The unit test infrastructure is SOOO HARD and it's the thing that's holding this whole process up.

**Comment by TurkeyMan** [comment]
What are we supposed to do about all these ancient test boxes?

**Comment by TurkeyMan** [comment]
The unit test fails because `hasElaborateCopyConstructor` is broken:
https://issues.dlang.org/show_bug.cgi?id=19902

**Comment by TurkeyMan** [comment]
@12345swordy `std::vector` is in, with a few known issues...

File: src/core/stdcpp/optional.d [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+ * Source:    $(DRUNTIMESRC core/stdcpp/optional.d)
+ */
+
+module core.stdcpp.optional;
+
+import core.stdcpp.exception : exception;
+
+version (CppRuntime_DigitalMars)
+{
+    pragma(msg, "std::optional not supported by DMC");
```
Comment (thewilsonator) [inline src/core/stdcpp/optional.d]: I think this should be a generic "DMC does not support C++11 or newer" generated by the compiler. i.e. if one passes `-extern-std=c++1[1|4|7] -m32`.

Comment (TurkeyMan) [reply src/core/stdcpp/optional.d]: DMC does have some sparse C++11 features.

Comment (thewilsonator) [reply src/core/stdcpp/optional.d]: Urgh, carry on then.

**Comment by TurkeyMan** [comment]
Buildkite fails... that other PR really needs attention ;)

**Comment by Geod24** [comment]
No it did not. And now master is broken. Once again.

**Comment by AndrejMitrovic** [comment]
Isn't it possible to make a github rule preventing merges until the PR is green? I also don't understand the lax attitude of merging every PR under the sun.

**Pull request merged**